### ISSUE TITLE
watchQuery causes TypeError: Cannot assign to read only property 'data' of object 

### ISSUE THREAD

**Issue by JohanHeyvaert** [apollographql/apollo-client#7507]
I don't know if I'm encountering a bug or if I'm doing something wrong. Every **first time** (after page load) my GQL query is executed, I get:
> TypeError: Cannot assign to read only property 'data' of object

Subsequent queries (until the page is reloaded) are working correctly! 
My code doesn't try to alter the received data, the error occurs while fetching the data, after a succesful reply of the server.

Here is the stack trace: 

> TypeError: Cannot assign to read only property 'data' of object
> at QueryInfo.js:210
> at perform (inMemoryCache.js:197)
> at InMemoryCache.push.kifG.InMemoryCache.performTransaction (inMemoryCache.js:214)
> at QueryInfo.push.FmZO.QueryInfo.markResult (QueryInfo.js:180)
> at QueryManager.js:520
> at asyncMap.js:11
> at new ZoneAwarePromise (zone-evergreen.js:960)
> at Object.next (asyncMap.js:11)
> at notifySubscription (Observable.js:135)
> at onNotify (Observable.js:179)

This is my code:

```
    const query = gql`
      query identify($number: String!) {
        identify(number: $number) {
          number
          lastName
        }
      }
    `;
    subscription = this.apollo
      .watchQuery<IdentifyResult>({
        query,
        variables: {
          number
        }
      })
      .valueChanges.subscribe(
        queryResult => {
          console.log('queryResult', queryResult);
        },
        error => {
          console.error(error);
        }
      );
```

In the AppModule I don't do anything special, I just add this provider:
```
    {
      provide: APOLLO_OPTIONS,
      useFactory: (httpLink: HttpLink) => {
        return {
          cache: new InMemoryCache(),
          link: httpLink.create({
            uri: 'http://localhost:60798/graphql/'
          })
        };
      },
      deps: [HttpLink]
    }
```

The error does not occur when I disable caching. 
```
          .watchQuery<Citizen>({
            query: this.myQuery,
            variables: {
              nationalNumber: action.nationalNumber
            },
            fetchPolicy: 'no-cache'
          })
```

I am using these libraries:
> "@apollo/client": "^3.3.6",
> "apollo-angular": "^2.1.0",
> "graphql": "^15.4.0",

Your help is much appreciated!

**Comment by benjamn** [comment]
@JohanHeyvaert That stack trace suggests the GraphQL `result` object in `result.data = ...` has been frozen, e.g. with `Object.freeze`. I don't think that's something Apollo Client is doing (though I'd be glad to fix it if I'm wrong about that). `InMemoryCache` does freeze the `result.data` object in development, but `result` itself shouldn't be frozen, and the `result.data` property should just be a normal (writable) property.

Can you try working backwards through the stack trace (in a debugger) to figure out where the freezing is happening? Or create a [reproduction](https://github.com/apollographql/react-apollo-error-template) so we can help you debug further?

**Comment by JohanHeyvaert** [comment]
Hi Ben,

Thank you for your response.

Weird that this isn't a known issue, as (I think) I am not doing anything special. Maybe it's related with the data returned by the backend?
Anyway, here is the result of the backwards debugging. I hope this helps to find the culprit.

Just before the exception is raised:
![image](https://user-images.githubusercontent.com/13747250/103637615-09798600-4f4c-11eb-8bf7-998d0df1008e.png)

Last line before the exception is triggered:
![image](https://user-images.githubusercontent.com/13747250/103637816-53fb0280-4f4c-11eb-8ae8-21b132df4e16.png)

Contents of data (same as optimisticData):
![image](https://user-images.githubusercontent.com/13747250/103638150-cbc92d00-4f4c-11eb-9949-2f201618d1cd.png)

Just before the finally block of inMemoryCache - performTransaction:
![image](https://user-images.githubusercontent.com/13747250/103638520-501bb000-4f4d-11eb-9ba7-db77aadece76.png)

performTransaction was called by QueryInfo - markResult:
![image](https://user-images.githubusercontent.com/13747250/103640334-1bf5be80-4f50-11eb-9f4e-02b983935846.png)

markResult was called by QueryManager - getResultsFromLink
![image](https://user-images.githubusercontent.com/13747250/103640389-34fe6f80-4f50-11eb-8654-2e0dbd55e4c5.png)

Contents of queryInfo:
![image](https://user-images.githubusercontent.com/13747250/103640654-b1914e00-4f50-11eb-8e89-baf0a335a929.png)


Thanks again for your help!

**Comment by intellix** [comment]
Related from the past: https://github.com/apollographql/apollo-client/issues/1909
This is stinging me again badly. Migrated to ApolloClient 3 and now getting alsorts of these errors throughout at runtime. Life is pain

**Comment by martin-braun** [comment]
Yes, I'm using @apollo/react-hooks 4.0.0 and my objects I get from my API are immutable. I don't want that, we need a way to disable it.

**Comment by JohanHeyvaert** [comment]
Any update on this? Just updated to v3.3.8 but the problem is still there. How is not everyone experiencing this problem? Is the caching feature no longer supported?

**Comment by martin-braun** [comment]
I had the same issue and simply disabled cache to bypass this limitation. All I wanted to do is modify and extend my results without making a deep copy. It's not the finest move, but the most convenient. I had to disable cache to make it work. 

@intellix, you pointed to that related issue in the past that got fixed in the PR #3883 which states: 

> This restriction was originally put in place to reserve the possibility of potentially returning direct cache references in query results.

This was on a time where no caching was possible, so the restriction got removed in that PR. Now we got cached results, so the restriction was added back for cached objects. So instead of saying that this is an issue, the question is if this is an issue at all?

Maybe @hwillson can step in place and give an explanation why results from the cache need to be deep frozen at all? It makes no sense to me and I guess it makes no sense to you, either. I should be able to modify an object that was returned from the cache or API every time I get it from the ApolloClient.

**Comment by JohanHeyvaert** [comment]
But the point is that I am not modifying the result at all. Just fetching the data causes the error. The data retrieval only works without error when disabling the cache. Cf. the above debugging information: all these actions and the eventual error are just triggered by calling apollo.watchQuery(). The fetch method causes the same error by the way.

That's why I don't understand why not everyone else is having this issue.

**Comment by martin-braun** [comment]
> But the point is that I am not modifying the result at all. Just fetching the data causes the error. The data retrieval only works without error when disabling the cache. Cf. the above debugging information: all these actions and the eventual error are just triggered by calling apollo.watchQuery(). The fetch method causes the same error by the way.
> 
> That's why I don't understand why not everyone else is having this issue.

Sorry, I should've read your issue, closely. You are right, this behavior is wrong. Well I couldn't pin point the source of the issue, but I can see that [result.data is set in the watchers code](https://github.com/apollographql/apollo-client/blob/v3.3.6/src/core/QueryInfo.ts#L392) a lot of times. I cannot see it being modified, so I doubt it's, because the `data` object is frozen. Instead, it looks like the _data_ property is marked read-only. 🤔

**Comment by JohanHeyvaert** [comment]
All good now!
